### PR TITLE
[AC-7729] API endpoint to reserve office hours

### PR DESCRIPTION
**Pull request by jonkiparsky**
https://masschallenge.atlassian.net/browse/AC-7729

### CODE REVIEW

**Review by iankigen — CHANGES_REQUESTED** [review]

**Review by geofrocker — CHANGES_REQUESTED** [review]
### A few issues for clean up
While on clean db, I did the following:
- I schedule office hours on behalf of this [mentor](http://localhost:8181/admin/simpleuser/user/1032/masquerade/)
- I masqueraded as this [finalist](http://localhost:8181/admin/simpleuser/user/13884/masquerade/) and successfully reserved an open office hour ✅
- I got the expected response but the startup field was empty yet the user has an associated startup ❌ ie
 ```javascript
"timecard_info": {
        "finalist_first_name": "User-First-13884",
        "finalist_last_name": "User-Last-13884",
        "topics": "",
        "startup": ""
    }
```
- I got an expected error message when I tried to reserve a non-existent office hour ie `No such office hour exists.` ✅
- I got an expected error message when I tried to reserve an already office hour ie `That session has already been reserved` ✅
- I got an expected error message when I tried to use a non-existent user ie `No such user exists` ✅
- I got an expected error message when I tried to use a non-existent startup ie `No such startup exists` ✅
- Startup id is required by the endpoint but not used for anything apart from checking that the startup exists ❌
- A user can pass any startup id that exists in the database and make a successful API call ❌

**Comment by jonkiparsky** [comment]
@geofrocker The "startup" field gets the startup that the user has selected for this office hour. It does not attempt to find a startup for the user. This is intentional, since the code will not have the information needed to make a sensible decision in this case.

**Comment by jonkiparsky** [comment]
@iankigen can you give more specific details on what you'd like the success header and detail to be?

File: web/impact/templates/emails/reserve_office_hour_email_to_mentor.html [hunk 1]
```diff
@@ -0,0 +1,19 @@
+Hi {{ recipient.first_name }},
+
+Your office hour session at {{ office_hour_date_time | date:"h:m" }} on
+{{ office_hour_date_time | date }} (office_hour_date_time : date:"%Z" }}
```
[the remark this answers is not in the document]
Comment (jonkiparsky) [reply web/impact/templates/emails/reserve_office_hour_email_to_mentor.html]: Please be more specific

**Comment by iankigen** [comment]
```
{
header: Office Hours session reserved on {date} (in  "day, date month, year" format if possible)
detail: with {mentor}, {mentor title} at {mentor company}
}
```
@jonkiparsky I think this format conforms to the mockup details.

File: web/impact/templates/emails/reserve_office_hour_email_to_mentor.html (hunk 1 above)
[the remark this answers is not in the document]
Comment (iankigen) [reply web/impact/templates/emails/reserve_office_hour_email_to_mentor.html]: we have a syntax error ln@4 `(office_hour_date_time : date:"%Z" }}
`

File: web/impact/impact/v1/views/reserve_office_hour_view.py [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+        if self.office_hour.finalist is not None:
+            self.fail(self.OFFICE_HOUR_ALREADY_RESERVED)
+            return False
+        self._update_office_hour_data()
+        self._send_confirmation_emails()
+        self._succeed()
+        return True
+
+    def _update_office_hour_data(self):
+        self.office_hour.finalist = self.target_user
```
[the remark this answers is not in the document]
Comment (jonkiparsky) [reply web/impact/impact/v1/views/reserve_office_hour_view.py:128]: 👍

File: web/impact/impact/v1/views/reserve_office_hour_view.py [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+        self.office_hour.description = self.message
+        self.office_hour.save()
```
[the remark this answers is not in the document]
Comment (jonkiparsky) [reply web/impact/impact/v1/views/reserve_office_hour_view.py:131]: 👍

File: web/impact/impact/v1/views/reserve_office_hour_view.py [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+
+    def _response(self):
+        return Response({
+            'success': self.success,
+            'header': self.header,
+            'detail': self.detail,
+            'timecard_info': self.timecard_info})
+
+
+def _variable_name(field_name):
```
[the remark this answers is not in the document]
Comment (jonkiparsky) [reply web/impact/impact/v1/views/reserve_office_hour_view.py]: 👍

**Review by iankigen — APPROVED** [review]

File: web/impact/impact/permissions/v1_api_permissions.py [hunk 5]
```diff
@@ -62,7 +62,7 @@ def has_permission(self, request, view):
                 is_expert(request.user))
 
 
-class ReserveOfficeHourPermission(V1APIPermissions):
+class ReserveOfficeHourPermission(IsAuthenticated):
     def has_permission(self, request, view):
         return (super().has_permission(request, view) or
```
Comment (iankigen) [inline web/impact/impact/permissions/v1_api_permissions.py]: can we check for is_employee user status?

**Review by iankigen — CHANGES_REQUESTED** [review]

**Review by iankigen — APPROVED** [review]
Code looks good

**Comment by geofrocker** [comment]
Looks good ✅

**Review by geofrocker — APPROVED** [review]

**Pull request merged**